mpcstats_lib.py

```diff
@@ -213,33 +228,59 @@ def _(i):
 
     return highest
 
-# LATER
+
+def variance(data: list[sint]):
+    return _variance(data, True)
+
+
+def linear_regression(xs: list[sint], ys: list[sint]):
+    # zip(xs, ys) is a list of data points
+
+    xs = Array.create_from(xs)
+    ys = Array.create_from(ys)
+
+    # calculate slope
+    covar = covariance(xs, ys)
+    var = variance(xs)
+    slope = covar / var
+
+    # calculate intercept
+    x_mean = _mean(xs)[0]
+    y_mean = _mean(ys)[0]
+    intercept = y_mean - slope * x_mean
+
+    res = sfix.Array(2)
+    res.assign([slope, intercept])
+    return res
+
 
 def harmonic_mean(data: list[sint]):
-    # TODO: implement harmonic_mean
-    raise NotImplementedError
+    eff_size = sum(if_else(n != MAGIC_NUMBER, 1, 0) for n in data)
+    eff_inv_total = sum(if_else(n != MAGIC_NUMBER, 1/n, 0) for n in data)
+    eff_inv_mean = eff_inv_total / eff_size
+    result = 1 / eff_inv_mean
 
+    # mpspdz doesn't raise division-by-zero error when data contains zeros
+    # whereas statistics.harmonic_mean returns zero if data contain zeros
+    # this aligns the behavior with the python counterpart
+    num_zeros = 0
+    for x in data:
+        num_zeros += (x == 0)
+    result *= (num_zeros == 0)
 
-def pstdev(data: list[sint]):
-    # TODO: implement pstdev
-    raise NotImplementedError
+    return result
 
 
 def pvariance(data: list[sint]):
-    # TODO: implement pvariance
-    raise NotImplementedError
+    return _variance(data, False)
 
 
-def stdev(data: list[sint]):
-    # TODO: implement stdev
-    raise NotImplementedError
-
+def pstdev(data: list[sint]):
+    pvar = _variance(data, False)
+    return sqrt(sfix(pvar))
```

Comment:
if `_variance` returned `sfix`, we could remove `sfix` here
```suggestion
    return sqrt(pvar)
```

Reply:
Sure. Changed.